### ISSUE TITLE
error 103: application.icon.image152x152 is an unexpected element/attribute

### ISSUE THREAD

**Issue by cybomon** [airsdk/Adobe-Runtime-Support#2544]
Using AIR 50.2.1

Submitting IPA to app loader fails with missing icon files of 152 px and 1024 px. Adding these dimensions to the app xml under the icon node produces an error from adt:

%appxml%.xml(44): error 103: application.icon.image152x152 is an unexpected element/attribute

**Comment by cybomon** [comment]
Hi Andrew, I figured out the adt error, there was a duplicate entry of that <image152ximage152> node. Yes, I'm using the asset.car file but thank you for linking that script. I've been using xcode and it's a pain. I'm still getting failure through the app loader when submitting the IPA. For some reason adt isn't packaging the icon files I have in the app xml. It doesn't throw an error either though. But when I look in the IPA they aren't there. I've done a compare with another app that went through and the app xml is almost the exact same except for the asset filenames and their paths. Could it be the asset.car I produced isn't kosher or something?

**Comment by cybomon** [comment]
In the info.plist in the IPA I do see this:

<key>CFBundleIconFiles</key>
                <array>
                    <string>Icon-Iphone-180.png</string>
                    <string>Icon-Ipad-152.png</string>

Which isn't the name I provided for the 152.png resource via adt, I see this in the root of the IPA so it must come from the Assets.car? However I don't see anything for 1024 px. So the plist references a valid 152 px icon that is packaged in the root. So what's with this error? Maybe it just doesn't like the assets.car?

**Comment by cybomon** [comment]
@ajwfrost I tried out the https://github.com/distriqt/AIR-ImageScripts, it makes the icon files but it seems to fail at asset.car creation: 
- Assets.car: AppIcon created
generate_appicon: AlternateIcon.appiconset doesn't exist - creating new icon set
 - Assets.car: AlternateIcon created
 - Assets.car: launch screen
 - Assets.car: actool
 - Assets.car: actool complete
cp: /Users/%user%/Documents/AIR-ImageScripts-master/.air-icon-generation-tmp/Assets.car: No such file or directory

And then the folder out structure has icons but no asset car:

![out](https://user-images.githubusercontent.com/5115176/227380360-18efcf08-c7d9-4f1a-be83-0f3c340d1c39.png)

**Comment by ajwfrost** [comment]
That may be happening if the actool is the wrong one .. if you try just running `xcrun actool` from the command line, does it work? You may need to do something like `sudo xcode-select -s /Applications/Xcode.app` or similar I think..

**Comment by cybomon** [comment]
xcrun actool didn't run, then I did the xcode command you gave and then xcrun actool did execute, silently. Then I tried the AIRimagescript tool again, in a new window. But the exact same outcome unfortunately.

**Comment by ajwfrost** [comment]
Just spotted the earlier comments too; the 1024x1024 icon shouldn't be included in the bundle itself as a separate file, but just in the assets.car file, so this png is stripped out by ADT. Similarly, all the filenames are changed (to match what Apple had originally suggested, although now apparently there are different filenames but they're not actually important, it's just the sizes that have to match).

From the errors then:

> ERROR ITMS-90023: "Missing required icon file. The bundle does not contain an app icon for iPad of exactly '152x152' pixels, in .png format for iOS versions >= 10.0

So your bundle should contain a PNG with the right size and this must also be referenced from the Info.plist file. This should happen as long as your app descriptor file has the image152x152 entry that points to the exact-size image..

> ERROR ITMS-90704: "Missing App Icon. An app icon measuring 1024 by 1024 pixels in PNG format must be included in the Asset Catalog

Which should be sorted out by the script to create the assets.car file from a 1024x1024 PNG file...


In that generate script, in the line where it runs actool, it redirects the output to null (something like `>/dev/null 2>/dev/null` or similar?) -> if you delete those parts and run it again, you should see any warnings from the tool itself?

**Comment by cybomon** [comment]
So I basically copied the exact app xml from the submitted app and it's asset structure, overwrote all the icon files with the same filenames just the new app icon. Exact same error. I'm still seeing that adt is not including all the icon files that are in the app xml for this app. So <image152x152>icons/Icon-App-76x76@2x.png</image152x152> is in the icon node in the app xml but it's not packaged in the IPA and adt doesn't throw an error. Here's the thing, it's not packaged in the app that submitted successfully to the app loader either. So it must be apple looks for those two inside assets.car and my assets.car file is broken some how. I'll try to troubleshoot the script with what you've provided. I wish we could just provide a one large icon and have the stores handle all this on the back end.

**Comment by cybomon** [comment]
@ajwfrost I removed the redirect to null part of that line and got this:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>com.apple.actool.errors</key>
	<array>
		<dict>
			<key>description</key>
			<string>Unknown argument '--include-all-app-icons'.</string>
		</dict>
	</array>
</dict>
</plist>
 - Assets.car: actool complete
cp: /Users/%username%/Documents/AIR-ImageScripts-master/.air-icon-generation-tmp/Assets.car: No such file or directory
========== COMPLETE ==============
```
Perhaps my version of xcode / tools is too old for this script? I'm using xcode ver 10.1.

**Comment by marchbold** [comment]
@cybomon Yes xcode 10 is very old we are currently expecting 14 for this script.

**Comment by cybomon** [comment]
Oh so that's why it sorry I should of included that with the air version. 
I was fiddling with xcode and generating the assets.car and I had the assets target membership selected for the app, but not for the test and uitest folders. I selected that and rebuilt. Then I noticed the asset.car was significantly larger. I became excited. It worked and passed the app loader! So it was asset.car being exported without those assets. Thanks for all your troubleshooting and guidance it helped out a lot! 

It would be awesome to use that appicon generate script however. I wonder if there's a way to just get that actool running, but then it might have dependencies too. 

I wonder if there's a way
